### PR TITLE
jsx supported, helpful classname added

### PR DESCRIPTION
**Pull request by gaoryrt**
- remove some hardcoded css
- welcomeMessage now support jsx

- new prop: helpLabel
- new prop: errorLabel
- new prop: historyAreaClassName
- new prop: historyCommandClassName

- rewrite prop: promptLabel
- remove prop: errorText

### CODE REVIEW

**Comment by gaoryrt** [comment]
trying to fullfill #210 
works well within my project, but not compatible with older versions

**Comment by linuswillner** [comment]
I'm unfortunately not in a position where I can review this at the moment. However, I've seen this and I'll get to it as soon as I can!

File: src/components/Terminal.jsx [hunk 1]
```diff
@@ -55,8 +55,6 @@ export default class Terminal extends React.Component {
     if (dangerMode) messageElement.innerHTML = message
     else messageElement.innerText = message
 
-    messageElement.style = 'margin: 0px; line-height: 21px;'
```
Comment (linuswillner) [inline src/components/Terminal.jsx]: This change is actually invalid since it compromises current functionality completely, but I was going to completely remove this functionality soon anyhow so we'll let it slide.

File: src/components/Terminal.jsx [hunk 2]
```diff
@@ -120,11 +118,12 @@ export default class Terminal extends React.Component {
   }
 
   showHelp () {
-    for (let c in this.state.commands) {
-      const cmdObj = this.state.commands[c]
-      const usage = cmdObj.usage ? ` - ${cmdObj.usage}` : ''
-      this.pushToStdout(`${c} - ${cmdObj.description}${usage}`)
-    }
+    const helpLabelFn = this.props.helpLabel
```
Comment (linuswillner) [inline src/components/Terminal.jsx]: When running this out of the box, I just get completely garbled output. The command I entered is also not displayed.

Garbled output

File: src/components/Terminal.jsx [hunk 3]
```diff
@@ -140,14 +139,10 @@ export default class Terminal extends React.Component {
   }
 
   getStdout () {
-    const lineStyle = {
```
Comment (linuswillner) [inline src/components/Terminal.jsx]: Why is `lineStyle` removed here, yet no new default style provided anywhere as far as I can see?

File: src/components/Terminal.jsx [hunk 4]
```diff
@@ -173,8 +168,16 @@ export default class Terminal extends React.Component {
       const args = input // ...and the rest can be used
 
       if (!this.props.noAutomaticStdout) {
-        if (!this.props.noHistory) this.pushToStdout(`${this.props.promptLabel || '$'} ${rawInput}`, rawInput)
-        else this.pushToStdout(`${this.props.promptLabel || '$'} ${rawInput}`)
+        this.pushToStdout([
```
Comment (linuswillner) [inline src/components/Terminal.jsx]: This block is indented wrong and has dangling commas, does not pass the linter phase.

File: src/components/Terminal.jsx [hunk 5]
```diff
@@ -184,7 +187,12 @@ export default class Terminal extends React.Component {
 
         const cmdObj = this.state.commands[command]
 
-        if (!cmdObj) this.pushToStdout(this.props.errorText ? this.props.errorText.replace(/\[command\]/gi, command) : `Command '${command}' not found!`)
+        if (!cmdObj) {
+          const errorLabelFn = this.props.errorLabel
```
Comment (linuswillner) [inline src/components/Terminal.jsx]: I also get completely garbled output here.

Garbled output

File: src/components/Terminal.jsx [hunk 6]
```diff
@@ -287,48 +295,48 @@ export default class Terminal extends React.Component {
       },
       prompt: {
         ...sourceStyles.prompt,
-        color: this.props.promptLabelColor || '#EE9C34'
```
Comment (linuswillner) [inline src/components/Terminal.jsx]: Where is the new default colour defined?

File: src/components/Terminal.jsx [hunk 7]
```diff
@@ -287,48 +295,48 @@ export default class Terminal extends React.Component {
...
+        color: this.props.promptLabelColor
       },
       input: {
         ...sourceStyles.input,
-        color: this.props.promptTextColor || '#F0BF81',
```
Comment (linuswillner) [inline src/components/Terminal.jsx]: Same as above, where is the new default colour defined?

File: src/components/Terminal.jsx [hunk 8]
```diff
@@ -287,48 +295,48 @@ export default class Terminal extends React.Component {
...
           >
-            <span
-              className={this.props.promptLabelClassName || null}
-              style={styles.prompt}
-            >
-              {this.props.promptLabel || '$'}
-            </span>
+            {
+              this.props.promptLabel
+                ? this.props.promptLabel()
```
Comment (linuswillner) [inline src/components/Terminal.jsx]: There's a smarter way to go about this. See my comment further down for nore details.

File: src/components/Terminal.jsx [hunk 9]
```diff
@@ -287,48 +295,48 @@ export default class Terminal extends React.Component {
...
             <input
               ref={this.terminalInput}
-              className={this.props.inputClassName || null}
+              className={this.props.inputClassName}
               name={'react-console-emulator__input'}
               style={styles.input}
               type={'text'}
               onKeyDown={this.handleInput}
               disabled={this.props.disableOnProcess && this.state.processing}
+              autoComplete="off"
```
Comment (linuswillner) [inline src/components/Terminal.jsx]: JSX keys should use singlequote.

File: src/utils/sourceStyles.js [hunk 10]
```diff
@@ -4,30 +4,24 @@ export const sourceStyles = {
     minHeight: '300px',
     maxWidth: '100%', // Fill parent before overflowing
     maxHeight: '100%', // Fill parent before overflowing
-    borderRadius: '5px',
```
Comment (linuswillner) [inline src/utils/sourceStyles.js]: Why were these removed?

File: src/utils/sourceStyles.js [hunk 11]
```diff
@@ -4,30 +4,24 @@ export const sourceStyles = {
...
-    overflow: 'auto',
-    cursor: 'text'
+    overflow: 'auto'
   },
   content: {
     padding: '20px',
     height: '100%',
-    fontSize: '15px'
```
Comment (linuswillner) [inline src/utils/sourceStyles.js]: Where is the new default font size defined?

File: src/utils/sourceStyles.js [hunk 12]
```diff
@@ -4,30 +4,24 @@ export const sourceStyles = {
...
   },
   inputArea: {
     display: 'inline-flex',
     width: '100%'
   },
   prompt: {
-    paddingTop: '3px'
```
Comment (linuswillner) [inline src/utils/sourceStyles.js]: Why was this removed?

File: src/utils/sourceStyles.js [hunk 13]
```diff
@@ -4,30 +4,24 @@ export const sourceStyles = {
...
+
   },
   input: {
     border: '0',
-    padding: '0 0 0 7px',
```
Comment (linuswillner) [inline src/utils/sourceStyles.js]: Why were the changes in this block made? To me, it only seems they completely break the terminal.

File: yarn.lock [hunk 14]
```diff
@@ -0,0 +1,8214 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (linuswillner) [inline yarn.lock]: I'd prefer to keep this repository without a Yarn lockfile.

File: src/utils/types.js [hunk 15]
```diff
@@ -23,9 +23,13 @@ export const types = {
...
     PropTypes.array,
-    PropTypes.string
+    PropTypes.string,
+    PropTypes.func
   ]),
-  promptLabel: PropTypes.string,
-  errorText: PropTypes.string,
-  commands: PropTypes.object.isRequired
+  commands: PropTypes.object.isRequired,
+  promptLabel: PropTypes.func,
```
Comment (linuswillner) [inline src/utils/types.js]: A better idea: Accept a node for this, instead of breaking existing code that uses the string variant.

**Comment by gaoryrt** [comment]
sorry for the bad code

File: src/components/Terminal.jsx (hunk 1 above)
Comment (gaoryrt) [reply src/components/Terminal.jsx]: One principle of this commit is use className and it's own css to style it instead of inline styles.

File: src/components/Terminal.jsx (hunk 2 above)
Comment (gaoryrt) [reply src/components/Terminal.jsx]: Sorry but this is a bug if you use tag 'dangerMode={true}', a react element would insert into a __html, rather than render directly.

File: src/components/Terminal.jsx (hunk 3 above)
Comment (gaoryrt) [reply src/components/Terminal.jsx]: Like I wrote above.
Classname-based style could be modified by both css and js, but inlineStyles can only be modified by js.
And, inlineStyles have higher priority than classname, people maynot change is as they want.

File: src/components/Terminal.jsx (hunk 5 above)
Comment (gaoryrt) [reply src/components/Terminal.jsx]: same inner-html bug

File: src/components/Terminal.jsx (hunk 6 above)
Comment (gaoryrt) [reply src/components/Terminal.jsx]: Maybe should put those default string in one place, easier to read and modify.

File: src/utils/sourceStyles.js (hunk 10 above)
Comment (gaoryrt) [reply src/utils/sourceStyles.js]: Inline styles is hard to rewrite, don't if it is not necessary.

File: src/utils/sourceStyles.js (hunk 13 above)
Comment (gaoryrt) [reply src/utils/sourceStyles.js]: I would like to write those css in a classname-based css files.
This default style do not fit my usage somehow so I deleted it.

File: src/utils/sourceStyles.js (hunk 12 above)
Comment (gaoryrt) [reply src/utils/sourceStyles.js]: redundant styles

**Comment by linuswillner** [comment]
Well, to be fair, it wasn’t really - your idea is excellent in spirit, but the execution was rather poor. I do actually agree with a lot of your points and have been locally working on a customisation rework, eventually planning to incorporate your idea in some form once the rework is complete. So while I’m not going to implement this in its current form, I do think your idea is still has a lot of merit and I’ll keep it in mind. Thanks for your time anyhow!
